### PR TITLE
Printhouse SFTP

### PR DESCRIPTION
**Pull request by stevenwilkin**
Need confirmation of paths for both files and if the trigger file needs any specific contents

### CODE REVIEW

File: app/services/upload_to_print_house.rb [hunk 1]
```diff
@@ -6,5 +8,17 @@ def initialize(csv)
...
   def call
+    upload_file('/Data.in/batch.csv', csv)
+    upload_file('/Trigger.in/trigger.txt', 'trigger')
+  end
+
+  private
+
+  def upload_file(path, contents)
+    Net::SFTP.start(ENV['SFTP_HOST'], ENV['SFTP_USER'], password: ENV['SFTP_PASSWORD']) do |sftp|
+      sftp.file.open(path, 'w') do |f|
```
Comment (gjvis) [inline app/services/upload_to_print_house.rb]: Need to use `sftp.upload` or the encoding goes screwy

**Comment by stevenwilkin** [comment]
We still need an end-to-end test to confirm the upload takes place.

I've played around with the following Python SFTP server which may be something we want to use:

https://github.com/rspivak/sftpserver

File: spec/services/upload_to_print_house_spec.rb [hunk 2]
```diff
@@ -2,8 +2,23 @@
...
+    allow(uploader).to receive(:upload_file)
+  end
+
+  it 'uploads the CSV' do
+    expect(uploader).to receive(:upload_file).with(csv_path, csv)
+    uploader.call
+  end
+
+  it 'uploads the trigger file' do
+    expect(uploader).to receive(:upload_file).with(trigger_path, trigger)
```
Comment (gjvis) [inline spec/services/upload_to_print_house_spec.rb]: I think you can use `have_received` in rspec 3+

**Comment by gjvis** [comment]
:+1:, and we'll sort out an end to end test once we've pulled the features across from the private repo

**Comment by stevenwilkin** [comment]
Updated!

**Comment by gjvis** [comment]
```
     .  o ..                  
     o . o o.o                
          ...oo               
            __[]__            
         __|_o_o_o\__         
         \""""""""""/         
          \. ..  . /          
     ^^^^^^^^^^^^^^^^^^^^ 
```

**Comment by stevenwilkin** [comment]
Love it!

File: app/services/upload_to_print_house.rb [hunk 3]
```diff
@@ -6,5 +9,16 @@ def initialize(csv)
   end
 
   def call
+    upload_file('/Data.in/batch.csv', csv)
```
Comment (andrewgarner) [inline app/services/upload_to_print_house.rb]: We know that filename should be:

```rb
"/Data.in/pensionwise_output_#{Date.today.strftime('%Y%m%d')}.csv"
```

Comment (stevenwilkin) [inline app/services/upload_to_print_house.rb]: Merci beaucoup!

File: app/services/upload_to_print_house.rb [hunk 4]
```diff
@@ -6,5 +9,16 @@ def initialize(csv)
...
+    upload_file('/Trigger.in/trigger.txt', 'trigger')
+  end
+
+  private
+
+  def upload_file(path, contents)
+    Net::SFTP.start(ENV['SFTP_HOST'], ENV['SFTP_USER'], password: ENV['SFTP_PASSWORD']) do |sftp|
+      io = StringIO.new(contents)
```
Comment (andrewgarner) [inline app/services/upload_to_print_house.rb]: Minor point, but, you don't need the SFTP connection open to create the `StringIO` instance.

```rb
io = StringIO.new(contents)

Net::SFTP.start(ENV['SFTP_HOST'], ENV['SFTP_USER'], password: ENV['SFTP_PASSWORD']) do |sftp|
  sftp.upload!(io, path)
end
```

Comment (gjvis) [inline app/services/upload_to_print_house.rb]: I'd argue the opposite direction - the `StringIO` isn't needed outside the block so should be scoped to the block only.

Comment (gjvis) [inline app/services/upload_to_print_house.rb]: Actually, I think @andrewgarner is right

Comment (andrewgarner) [inline app/services/upload_to_print_house.rb]: But then you're keeping the connection open so that you can do local processing.

**Comment by stevenwilkin** [comment]
Fixup commits added

**Comment by gjvis** [comment]
Looks good I think. We need to add some logging into this, but that can be a separate PR

**Comment by stevenwilkin** [comment]
Manually merged into master as this was part of a trio of PRs

**Pull request closed without merging**